modules/assim.batch/R/pda.emulator.R

```diff
@@ -86,76 +83,205 @@ pda.emulator <- function(settings, params.id=NULL, param.names=NULL, prior.id=NU
+    run.block = TRUE
+  }else{
+    run.block = FALSE
   }
-
-  ## Collect all likelihoods (Y)
-  # For now, just the likelihoods from the runs we just did. 
-  # *** TODO: Soon, need to add ability to retrieve them from previous runs, because we want to build the emulator from as many points as possible. The likelihoods are easy—they're being stored in BETY. But we need to know the parameter values associated with them too, and that will take a bit of doing.
- 
+  
+  
+  if(run.block){
+    ## Set up runs and write run configs for all proposed knots 
+    run.ids <- pda.init.run(settings, con, my.write.config, workflow.id, knots.params, 
+                            n=settings$assim.batch$n.knot, 
+                            run.names=paste0(settings$assim.batch$ensemble.id,".knot.",1:settings$assim.batch$n.knot))
+    
+    ## start model runs
+    start.model.runs(settings,settings$database$bety$write)
+    
+    ## Retrieve model outputs, calculate likelihoods (and store them in database)
+    LL.0 <- rep(NA, settings$assim.batch$n.knot)
+    model.out <- list()
+    
+    for(i in 1:settings$assim.batch$n.knot) {
+      ## read model outputs
+      model.out[[i]] <- pda.get.model.output(settings, run.ids[i], inputs)
+      
+      ## calculate likelihood
+      LL.0[i] <- pda.calc.llik(settings, con, model.out[[i]], run.ids[i], inputs, llik.fn)
+    }
+  } 
+  
+  
   ## if it is not specified, default to GPfit
   if(is.null(settings$assim.batch$GPpckg)) settings$assim.batch$GPpckg="GPfit"
-
-  if(settings$assim.batch$GPpckg=="GPfit"){
-    ## GPfit optimization routine assumes that inputs are in [0,1]
-    ## Instead of drawing from parameters, we draw from probabilities
-    X <- knots.probs[, prior.ind, drop=FALSE]
-
-    logger.info(paste0("Using 'GPfit' package for Gaussian Process Model fitting."))
-    require(GPfit)
-    ## Generate emulator on LL-probs
-    GPmodel <- GP_fit(X=X,Y=LL.X)
-    gp=GPmodel
+  
+  if(settings$assim.batch$GPpckg=="GPfit"){ # GPfit-if
+    
+    if(run.block){ 
+      
+      ## GPfit optimization routine assumes that inputs are in [0,1]
+      ## Instead of drawing from parameters, we draw from probabilities
+      X <- knots.probs[, prior.ind, drop=FALSE]
+      
+      LL.X <- cbind(X, LL.0)
+      
+      if(!is.null(settings$assim.batch$extension)){ 
+        # check whether another "round" of emulator requested
+        
+        # load original knots
+        load(settings$assim.batch$llik.path)
+        LL <- rbind(LL.X, LL)
+        
+      }else{ 
+        LL <- LL.X
+      }
+      
+      logger.info(paste0("Using 'GPfit' package for Gaussian Process Model fitting."))
+      require(GPfit)
+      ## Generate emulator on LL-probs
+      GPmodel <- GP_fit(X = LL[,-ncol(LL), drop=FALSE],
+                        Y = LL[,ncol(LL), drop=FALSE])
+      gp=GPmodel
+      
+    }else{
+      load(settings$assim.batch$emulator.path) # load previously built emulator to run a longer mcmc
```

Comment:
This isn't cause for changing anything in this pull request, but I think it might be a good idea to shift from using RData objects to using RDS objects (and associated functions). The disadvantage of the latter is that you can only save one object at a time in an RDS file, but the major advantage is that you can then load it into a variable, rather than silently into its previous name the way "load" does. Just an idea...